### DISCUSSION TITLE
Feature Request: more granular configuration options

### DISCUSSION THREAD

**Discussion by armanm** [duckdb/duckdb#5724] [Ideas]
I would like to use Duckdb in a tenanted environment. Therefore it will be useful to limit access to some constructs of Duckdb per user with more granular configuration flags that control the following aspects:

* limit external file (CSV and parquet files) access to specific paths
* enable/disable external file access from S3
* preload and limit the connection to a set of specified extensions
* limit max query result per query

Today you are able control to some of the above using the `enable_external_access` setting. It acts as an all or nothing flag. Ideally the following configs would allow for tailored limitations depending on who is going to execute queries:

```js
new duckdb.Database(":memory:", {
  "enable_s3_access": false,
  "files_path": "/Home/users/foo,/Home/users/bar",
  "enabled_modules": "parquet,excel",                     
  "max_results_per_query": "1000"                             
})
```

* `enable_s3_access` - would control if loading data from S3 is allowed
* `files_path` - similar to `file_search_path` but would lock Duckdb to files under one or more directories on the disk
* `enabled_modules` - allow controlling which set of modules can be enabled in a particular connection
* `max_results_per_query` - controls the max number of records a user is allowed to return with each query to avoid requests that take a long time to finish

Thanks in advance

**Comment by earwig** [comment]
In the same vein, we would find it useful to have a way to disable replacement scans without turning off `enable_external_access`.

The pandas replacement scans feature in the Python client has lead to hard-to-debug issues with misspelled table names and data frames we didn't realize were in-scope.

**Comment by karlseguin** [comment]
I would leverage all of the feature requests made so far.  

An additional security feature I'd like to see, which is probably in a different league implementation-wise,  is something like sqlite's authorization callbacks.

**Reply by karlseguin** [reply]
Something perhaps more realistic than sqlite's authorization callback would be the ability to put a specific connection in a readonly mode. This is different than the current `access_mode`, which is database-wide. The goal would be to allow some connections to read-write and others to be readonly (within the same process).

**Reply by Mause** [reply]
This is already the case - `access_mode` is already connection specific

**Reply by karlseguin** [reply]
I don't think I explained myself very well. This would be like calling `set session characteristics as transaction read only;` in PostgreSQL. Clearly this isn't how `access_mode` behaves.

In PostgreSQL:
```
set session characteristics as transaction read only;
SET
test=# create table x (id varchar);
ERROR:  cannot execute CREATE TABLE in a read-only transaction
```

In DuckDB:
```
./duckdb test.db
v0.8.0 e8e4cea5ec
Enter ".help" for usage hints.
D set access_mode=read_only;
D create table x (id varchar);
D insert into x(id) values ('1');
D select * from x;
┌─────────┐
│   id    │
│ varchar │
├─────────┤
│ 1       │
└─────────┘
```

Another way to think about it is that it would be a way of creating readonly users.

**Reply by Mause** [reply]
Access mode can't be set after the database is opened (this should probably error instead). Try passing the -readonly flag to the CLI

**Reply by karlseguin** [reply]
Thanks, I got it working and I think this'll achieve what I was after. My only complaint then would be the documentation :)

**Reply by karlseguin** [reply]
I tested this more and it appears to have some limitations. Namely, the readonly connections don't see what the read_write connections are doing (within the same process). I'm not sure this is the intended behaviour or a bug.

To be clear, this is within the same process, using the C-API. I'm not sure how to simulate this using just the CLI while staying within the same process.

This works, because the `test` table is created before the read_only database/connection is created:
```
rw_db = duckdb_open_ext("db.duckdb");
rw_conn = duckdb_connect(rw_db);
duckdb_query(rw_conn, "create table test (id int)");

ro_db = duckdb_open_ext("db.duckdb", "access_mode=read_only");
ro_conn = duckdb_connect(ro_db);
duckdb_query(ro_conn, "select * from test");
```

This doesn't work, because the read only database/connection is opened before `test` is created (you get a table does not exist error):
```
rw_db = duckdb_open_ext("db.duckdb");
rw_conn = duckdb_connect(rw_db);
ro_db = duckdb_open_ext("db.duckdb", "access_mode=read_only");
ro_conn = duckdb_connect(ro_db);

duckdb_query(rw_conn, "create table test (id int)");
duckdb_query(ro_conn, "select * from test");
```

The use-case here is an app that allows a group of users (admins) to write to the DB, and another group of users to query the same database.

**Reply by Mause** [reply]
Could you provide some code that shows the issue? The code you provided isn't using the C API/doesn't compile, but I just wrote a test myself using pthread and it seemed to work fine

**Reply by karlseguin** [reply]
I'm not a C programmer, but this reproduces the issue when run the first time:

```c
#include <stdio.h>
#include "duckdb.h"

int main() {
  duckdb_config rw_config;
  duckdb_database rw_db;
  char *err;

  if (duckdb_create_config(&rw_config) == DuckDBError) {
    printf("failed to create rw config\n");
    exit(1);
  }

  // I thought this was the issue, but it turns out this code also fails when this is removed
  duckdb_set_config(rw_config, "access_mode", "READ_WRITE");

  if (duckdb_open_ext("/tmp/test.duckdb", &rw_db, rw_config, &err) == DuckDBError) {
    printf("failed to open rw db: %s\n", err);
    exit(1);
  }

  duckdb_config ro_config;
  duckdb_database ro_db;
  if (duckdb_create_config(&ro_config) == DuckDBError) {
    printf("failed to create ro config\n");
    exit(1);
  }
  duckdb_set_config(rw_config, "access_mode", "READ_ONLY");

  if (duckdb_open_ext("/tmp/test.duckdb", &ro_db, ro_config, &err) == DuckDBError) {
    exit(1);
  }

  duckdb_connection rw_conn;
  if (duckdb_connect(rw_db, &rw_conn) == DuckDBError) {
    exit(2);
  }

  duckdb_connection ro_conn;
  if (duckdb_connect(ro_db, &ro_conn) == DuckDBError) {
    exit(3);
  }

  if (duckdb_query(rw_conn, "create table if not exists x(id int)", NULL) == DuckDBError) {
    exit(4);
  }

  // This is the code that fails which I was hoping would not
  duckdb_result result;
  if (duckdb_query(ro_conn, "select * from x", &result) == DuckDBError) {
    printf("Select failure: %s\n", duckdb_result_error(&result));
  }

  duckdb_destroy_result(&result);
  duckdb_disconnect(&rw_conn);
  duckdb_disconnect(&ro_conn);
  duckdb_close(&rw_db);
  duckdb_close(&ro_db);
}
```

First time this is run, you'll get:

> Select failure: Catalog Error: Table with name x does not exist!

Run it a 2nd time, and it works fine since table "x" now exists for both to see.

I thought this was related to the access_mode, i.e. `duckdb_set_config(rw_config, "access_mode", "READ_ONLY");`  but now I realize this error happens even if you remove this line. I couldn't find anything in the documentation that says you can/cannot open 2 databases to the same file from within a process. I guess you can, but they won't see each other's work. I realize this is probably unusual, but I was hoping it would be one way to have read_write and read_only connections from within a process. Can't figure out any other way to do it.

**Reply by Mause** [reply]
My initial example (and assumptions around how to do this) were slightly off, here's a working example which I hope helps (you'll notice the shared database instance but two connections to it)

<details><summary>Code example</summary>
<p>

```c++
#include "src/include/duckdb.h"

#include <pthread.h>
#include <stdio.h>
#include <unistd.h>

#define OK(RET)                                                                                                        \
	if ((RET) != 0) {                                                                                                  \
		printf(#RET ": %d\n", RET);                                                                                    \
	}

duckdb_result *run(duckdb_connection rw_conn, const char *query) {
	duckdb_result *result = static_cast<duckdb_result *>(malloc(sizeof(duckdb_result)));
	if (duckdb_query(rw_conn, query, result)) {
		printf("\"%s\" failure: %s\n", query, duckdb_result_error(result));
	}
	return result;
}

void *worker(void *rw_db) {
	duckdb_connection ro_conn;
	OK(duckdb_connect(static_cast<duckdb_database>(rw_db), &ro_conn));
	printf("connected in thread, will wait for data\n");
	sleep(1);
	printf("querying table in thread\n");

	duckdb_result *result = run(ro_conn, "select * from x");
	int *data = static_cast<int *>(duckdb_column_data(result, 0));
	printf("result: %d\n", *data);
	duckdb_destroy_result(result);

	duckdb_disconnect(&ro_conn);

	return nullptr;
}

int main() {
	duckdb_database rw_db;
	const char *path = "/tmp/test.duckdb";
	unlink(path);

	OK(duckdb_open(path, &rw_db));

	duckdb_connection rw_conn;
	OK(duckdb_connect(rw_db, &rw_conn));

	pthread_t thread_obj;
	pthread_create(&thread_obj, nullptr, &worker, rw_db);

	duckdb_destroy_result(run(rw_conn, "create table if not exists x(id int)"));
	duckdb_destroy_result(run(rw_conn, "insert into x values (42)"));
	printf("created table\n");

	pthread_join(thread_obj, nullptr);

	duckdb_disconnect(&rw_conn);
	duckdb_close(&rw_db);
}
```


</p>
</details> 

Hope this helps!

**Reply by Mause** [reply]
Though I realise now while this works, it doesn't fulfill your initial requirements of a single RW thread and another RO thread. I might have to leave the question of whether that is possible after all to @Mytherin

**Comment by karlseguin** [comment]
As an alternative to `files_path` it might be useful to have a `file_root` which not only locks access to the specific path, but also serves as the root for relative paths.

Give a `file_root` of `/files/tenant/2`  a `select from`  with either a target of `/files/x.json` or `files/x.json` would result in loading data from `/data/tenant/2/files/x.json`.  And, of course, the caller **could not** break out of the `/files/tenant/2/` root (via a target of `../etc/passwd` for example)

**Reply by Tishj** [reply]
Symbolic links might mess with that, but I like the idea
